### ISSUE TITLE
Split: Train Test Validation 

### ISSUE THREAD

**Issue by peterdudfield** [openclimatefix/nowcasting_utils#16]
We need a good way to split up data into train/validation/data sets. 

Would like a function that takes a a list of timestamps (List[pd.Timestamp]) and splits according to a few methods. Note we are delaying with timeseries data so we have to be a little careful about sampling. This is so as not to give samples are that are too similar in different sets. Because Solar generation is zero in the night, we can in general make sure we split data around midnight. 

Function inputs:
- List[pd.Timestamp]: all time series
- method:str, the method to use
- train_test_validation_split: List[int] - This gives the ratio of data that should be put in the different sets

Methods:
- Day random - Divide the dataset into day sgroups, and randomly take samples according the 'train_test_validation_split'
- Week random  - Divide the dataset into weeks groups, and randomly take samples according the 'train_test_validation_split'
- Full - return full datestet to all new datasets. This might be useful for unit-testing with a small amount of data
- Any others?

Return  dict[str:List[pd.Timestamp]] - return three datasets. The keys should be 'train','validation' and 'test'

**Comment by jacobbieker** [comment]
Possibly a yearly split? Like train on all 2020 and test on 2021?

**Comment by peterdudfield** [comment]
yea perhaps we could have 
- random year - take random years to training set
- specific year - takes an input which tells which years to go into which datasets

**Comment by jacobbieker** [comment]
Yeah, I think would be a good option for splits too

**Comment by JackKelly** [comment]
Some existing notes in this issue: https://github.com/openclimatefix/nowcasting_dataset/issues/7  :)

**Comment by JackKelly** [comment]
Great work, thinking about this!

A few thoughts:

* Should `nowcasting_dataset` be responsible for splitting the dataset?  The idea would be that `nowcasting_dataset` would output batches into different folders for each split?  This is nice because it simplifies the training code, and (almost) guarantees that all the experiments we conduct on a particular version of the dataset will use the exact same split (hence the results should be comparable).
* We can ignore this complexity for now, but further down the line we may also want to split by PV system ID (so, for example, the test and validation sets would only use PV systems that the model hadn't seen during training).
* When our dataset spans many years, I'd perhaps be in favour of a super-simple split, like @jacobbieker suggests:  Train on 2015 to 2020; test on 2021.  Simple splits are good when we're presenting our results to people who may never have come across the concept of splitting datasets :)   And this is also nice when we're comparing against ESO's existing forecasts, which I think have only have been running for a year or two.
* But, before we have downloaded multiple years of data, another simple approach would be something like "take every 5th day for the validation set" (or every 5th week).  I think I remember seeing this sort of approach in one of the recent DeepMind papers.
* Which reminds me:  We should maybe do a super-quick review of how some recent nowcasting papers have split their datasets!

**Comment by JackKelly** [comment]
Sorry, one more thought:  Splitting timeseries data is surprisingly fiddly and error-prone, which is another argument in favour of keeping the split as simple as possible, and implementing as few options as possible :)   (I'm always on a mission to see how _little_ code we can write!)

**Comment by peterdudfield** [comment]
Yea I agree, the splitting data by every 5th day should be the first thing to do. 

Could move this back to dataset, TBH not sure why I wanted it here

Yea split by PV system- can be a later thing

Yea the split by year, and saying what year, should be prioty number 2

**Comment by JackKelly** [comment]
Cool, I'll transfer this issue back to `nowcasting_dataset`, and close the old issue as a duplicate of this one :)
